src/librustc_metadata/locator.rs

```diff
@@ -656,14 +656,34 @@ impl<'a> CrateLocator<'a> {
         dylibs: FxHashMap<PathBuf, PathKind>,
     ) -> Option<(Svh, Library)> {
         let mut slot = None;
+        // Order here matters, rmeta should come first. See comment in
+        // `extract_one` below.
         let source = CrateSource {
-            rlib: self.extract_one(rlibs, CrateFlavor::Rlib, &mut slot),
             rmeta: self.extract_one(rmetas, CrateFlavor::Rmeta, &mut slot),
+            rlib: self.extract_one(rlibs, CrateFlavor::Rlib, &mut slot),
             dylib: self.extract_one(dylibs, CrateFlavor::Dylib, &mut slot),
         };
         slot.map(|(svh, metadata)| (svh, Library { source, metadata }))
     }
 
+    fn needs_crate_flavor(&self, flavor: CrateFlavor) -> bool {
+        if flavor == CrateFlavor::Dylib && self.is_proc_macro == Some(true) {
+            return true;
+        }
+
+        let is_rlib = self.sess.crate_types.borrow() == &[config::CrateType::Rlib];
```

Comment:
```suggestion
        let is_rlib = self.sess.crate_types.borrow().all(|crate_type| crate_type == config::CrateType::Rlib);
```
Technically `rustc --crate-type=rlib --crate-type=rlib lib.rs` is legal and I don't see `crate_types` deduplicated anywhere.

Reply:
Crazy! But indeed true.

I've done a bit of an audit of where we use crate_types and a few interesting things turned up:

In save analysis, we seem to write out the output path, but it seems not obviously wrong at least: https://github.com/Mark-Simulacrum/rust/blob/0d0006b76bfd8fa9b3df18986b1333dfa46984f3/src/librustc_save_analysis/lib.rs#L90

Interestingly in https://github.com/Mark-Simulacrum/rust/blob/0d0006b76bfd8fa9b3df18986b1333dfa46984f3/src/librustc_interface/passes.rs#L371 it seems that `--crate-type proc-macro --crate-type proc-macro` will actually error today.

And in https://github.com/Mark-Simulacrum/rust/blob/0d0006b76bfd8fa9b3df18986b1333dfa46984f3/src/librustc_interface/passes.rs#L480 we will generate duplicate output filenames...

https://github.com/Mark-Simulacrum/rust/blob/0d0006b76bfd8fa9b3df18986b1333dfa46984f3/src/librustc_codegen_ssa/back/write.rs#L767 will consider `--crate-type rlib --crate-type rlib` to not be an rlib, which means you can get full LTO even in rlibs.

(possibly some more)

Generally speaking it looks like dedup'ing these is a good idea so I've filed https://github.com/rust-lang/rust/pull/68403